aggregator/src/datastore.rs

```diff
@@ -2885,6 +3070,155 @@ ORDER BY id DESC
+    pub async fn delete_old_client_reports(
+        &self,
+        task_id: &TaskId,
+        oldest_allowed_report_timestamp: Time,
+    ) -> Result<(), Error> {
+        let stmt = self
+            .tx
+            .prepare_cached(
+                "DELETE FROM client_reports
+                WHERE task_id = (SELECT id FROM tasks WHERE task_id = $1) AND client_timestamp < $2
+                AND NOT EXISTS(
+                    SELECT FROM report_aggregations
+                    WHERE report_aggregations.client_report_id = client_reports.id)",
+            )
+            .await?;
+        self.tx
+            .execute(
+                &stmt,
+                &[
+                    /* task_id */ &task_id.get_encoded(),
+                    /* client_timestamp */
+                    &oldest_allowed_report_timestamp.as_naive_date_time()?,
+                ],
+            )
+            .await?;
+        Ok(())
+    }
+
+    /// Deletes old aggregation artifacts (aggregation jobs/report aggregations/batch aggregations)
+    /// for a given task, that is, aggregation artifacts for which all associated client reports
+    /// have timestamps older than a given timestamp, and which are not included in any collection
+    /// artifacts.
+    ///
+    /// Only implemented for leader tasks with the time-interval query type.
+    ///
+    /// After calling this function, delete_old_client_reports must be called in the same
+    /// transaction to avoid re-aggregating client reports.
+    #[tracing::instrument(skip(self), err)]
+    pub async fn delete_old_aggregation_artifacts(
+        &self,
+        task_id: &TaskId,
+        oldest_allowed_report_timestamp: Time,
+    ) -> Result<(), Error> {
+        let stmt = self
+            .tx
+            .prepare_cached(
+                "WITH aggregation_jobs_to_delete AS (
+                    SELECT id FROM aggregation_jobs
+                    JOIN (
+                        SELECT
+                            report_aggregations.aggregation_job_id,
+                            MAX(client_reports.client_timestamp) AS max_timestamp
+                        FROM report_aggregations
+                        JOIN client_reports
+                            ON client_reports.id = report_aggregations.client_report_id
+                        GROUP BY report_aggregations.aggregation_job_id
+                    ) report_max_timestamps
+                        ON report_max_timestamps.aggregation_job_id = aggregation_jobs.id
+                    WHERE aggregation_jobs.task_id = (SELECT id FROM tasks WHERE task_id = $1)
+                      AND report_max_timestamps.max_timestamp < $2
+                      AND NOT EXISTS (
+                          SELECT id FROM collect_jobs
+                          WHERE aggregation_jobs.task_id = collect_jobs.task_id
+                            AND aggregation_jobs.batch_interval <@ collect_jobs.batch_interval)
+                ),
```

Comment:
Would this query be simpler if the aggregation jobs table had a column for the most recent timestamp of any report in the aggregation job? Then we would just need to check if that column's value is farther in the past than `oldest_allowed_report_timestamp`, right?

Reply:
It would, but this would incur a schema change & data migration. I'd also prefer to await denormalization until we know which queries are slow.

Reply:
Oh, yes, good point about backfills.